stats/opencensus/stats.go

```diff
@@ -0,0 +1,227 @@
+	// number of messages sent from side (client || server)
+	sentMsgs int64
+	// number of bytes sent (within each message) from side (client || server)
+	sentBytes int64
+	// number of messages received on side (client || server)
+	recvMsgs int64
+	// number of bytes received (within each message) received on side (client
+	// || server)
+	recvBytes int64
+
+	startTime time.Time // is this still the correct way to count latency?
+	method    string
+}
+
+// statsTagRPC creates a recording object to derive measurements from in the
+// context, scoping the recordings to per RPC Attempt client side (scope of the
+// context). It also populates the gRPC Metadata within the context with any
+// opencensus specific tags set by the application in the context, binary
+// encoded to send across the wire.
+func (csh *clientStatsHandler) statsTagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	d := &rpcData{
+		startTime: time.Now(),
+		method:    info.FullMethodName,
+	}
+	// Populate gRPC Metadata with OpenCensus tag map if set by application.
+	if tm := tag.FromContext(ctx); tm != nil {
+		ctx = stats.SetTags(ctx, tag.Encode(tm))
+	}
+	return setRPCData(ctx, d)
+}
+
+// statsTagRPC creates a recording object to derive measurements from in the
+// context, scoping the recordings to per RPC server side (scope of the
+// context). It also deserializes the opencensus tags set in the context's gRPC
+// Metadata, and adds a server method tag to the opencensus tags.
+func (ssh *serverStatsHandler) statsTagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	d := &rpcData{
+		startTime: time.Now(),
+		method:    info.FullMethodName,
+	}
+	if tagsBin := stats.Tags(ctx); tagsBin != nil {
+		if tags, err := tag.Decode(tagsBin); err != nil {
+			ctx = tag.NewContext(ctx, tags)
+		}
+	}
+	ctx, _ = tag.New(ctx, tag.Upsert(keyServerMethod, removeLeadingSlash(info.FullMethodName)))
+	return setRPCData(ctx, d)
+}
+
+func recordRPCData(ctx context.Context, s stats.RPCStats) {
+	switch st := s.(type) {
+	case *stats.InHeader, *stats.OutHeader, *stats.InTrailer, *stats.OutTrailer:
+		// Headers and Trailers are not relevant to the measures, as the
+		// measures concern number of messages and bytes for messages. This
+		// aligns with flow control.
+	case *stats.Begin:
+		recordDataBegin(ctx, st)
+	case *stats.OutPayload:
+		recordDataOutPayload(ctx, st)
+	case *stats.InPayload:
+		recordDataInPayload(ctx, st)
+	case *stats.End:
+		recordDataEnd(ctx, st)
+	default:
+		// Shouldn't happen. gRPC calls into stats handler, and will never not
+		// be one of the types above.
+	}
+}
+
+// recordDataBegin takes a measurement related to the RPC beginning,
+// client/server started RPCs dependent on the caller.
+func recordDataBegin(ctx context.Context, b *stats.Begin) {
+	d, ok := ctx.Value(rpcDataKey{}).(*rpcData)
+	if !ok {
+		// Shouldn't happen, as gRPC calls TagRPC which populates the rpcData in
+		// context.
+		return
+	}
```

Comment:
How about moving this into `recordRPCData` so it only needs to happen in one place?